### PR TITLE
Use machine config operator (MCO) to make kubelet conf change

### PR DESCRIPTION
**Pull request by praveenkumar**
This patch will use MCO for updating the dynamic resourcfile which
is added recently with openshift/machine-config-operator#2466.

### CODE REVIEW

**Comment by guillaumerose** [comment]
Great! I remember I tried couple of months ago without luck! 

/lgtm

**Comment by praveenkumar** [comment]
Tested generated bundle and it does work. `reserved=cpu=0m,memory=0Gi`
```
kubelet --config=/etc/kubernetes/kubelet.conf --bootstrap-kubeconfig=/etc/kubernetes/kubeconfig --kubeconfig=/var/lib/kubelet/kubeconfig --container-runtime=remote --container-runtime-endpoint=/var/run/crio/crio.sock --runtime-cgroups=/system.slice/crio.service --node-labels=node-role.kubernetes.io/master,node.openshift.io/os_id=rhcos --node-ip=192.168.126.11 --minimum-container-ttl-duration=6m0s --cloud-provider= --volume-plugin-dir=/etc/kubernetes/kubelet-plugins/volume/exec --register-with-taints=node-role.kubernetes.io/master=:NoSchedule --pod-infra-container-image=quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:1ecbe11e2366c3a497e50e7f2f03589ff7da1c67217b28ed38d2ffa5c8e10e3b --system-reserved=cpu=0m,memory=0Gi --v=2

```

**Comment by praveenkumar** [comment]
@cfergeau Have a look to this.

File: 99_master-dynamic-system-reserved-calc.yaml [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+    machineconfiguration.openshift.io/role: master
+  name: 99-remove-systemd-reserved-mem-cpu
+spec:
+  config:
+    ignition:
+      version: 3.2.0
+    storage:
+      files:
+      - contents:
+          source: data:text/plain;charset=utf-8;base64,IyEvYmluL2Jhc2gKCnNldCAtZQoKTk9ERV9TSVpFU19FTlY9JHtOT0RFX1NJWkVTX0VOVjotL2V0Yy9ub2RlLXNpemluZy5lbnZ9CgpmdW5jdGlvbiBkeW5hbWljX25vZGVfc2l6aW5nIHsKICAgIHJtIC1mICR7Tk9ERV9TSVpFU19FTlZ9CiAgICBlY2hvICJTWVNURU1fUkVTRVJWRURfTUVNT1JZPTFHaSIgPj4gJHtOT0RFX1NJWkVTX0VOVn0KICAgIGVjaG8gIlNZU1RFTV9SRVNFUlZFRF9DUFU9NTAwbSIgPj4gJHtOT0RFX1NJWkVTX0VOVn0KfQoKZnVuY3Rpb24gc3RhdGljX25vZGVfc2l6aW5nIHsKICAgIHJtIC1mICR7Tk9ERV9TSVpFU19FTlZ9CiAgICBlY2hvICJTWVNURU1fUkVTRVJWRURfTUVNT1JZPTBHaSIgPj4gJHtOT0RFX1NJWkVTX0VOVn0KICAgIGVjaG8gIlNZU1RFTV9SRVNFUlZFRF9DUFU9MG0iID4+ICR7Tk9ERV9TSVpFU19FTlZ9Cn0KCmlmIFsgJDEgPT0gInRydWUiIF07IHRoZW4KICAgIGR5bmFtaWNfbm9kZV9zaXppbmcKZWxpZiBbICQxID09ICJmYWxzZSIgXTsgdGhlbgogICAgc3RhdGljX25vZGVfc2l6aW5nCmVsc2UKICAgIGVjaG8gIlVucmVjb25nbml6ZWQgY29tbWFuZCBsaW5lIG9wdGlvbi4gVmFsaWQgb3B0aW9ucyBhcmUgXCJ0cnVlXCIgb3IgXCJmYWxzZVwiIgpmaQoK
```
Comment (cfergeau) [inline 99_master-dynamic-system-reserved-calc.yaml:14]: I don't know if there's an 'official' way of doing this, but can we store the real source file in git, and generate this file at build time?
In git: 
```
$ cat dynamic-system-reserved-calc.sh
#!/bin/bash

set -e

NODE_SIZES_ENV=${NODE_SIZES_ENV:-/etc/node-sizing.env}

function dynamic_node_sizing {
    rm -f ${NODE_SIZES_ENV}
    echo "SYSTEM_RESERVED_MEMORY=1Gi" >> ${NODE_SIZES_ENV}
    echo "SYSTEM_RESERVED_CPU=500m" >> ${NODE_SIZES_ENV}
}

function static_node_sizing {
    rm -f ${NODE_SIZES_ENV}
    echo "SYSTEM_RESERVED_MEMORY=0Gi" >> ${NODE_SIZES_ENV}
    echo "SYSTEM_RESERVED_CPU=0m" >> ${NODE_SIZES_ENV}
}

if [ $1 == "true" ]; then
    dynamic_node_sizing
elif [ $1 == "false" ]; then
    static_node_sizing
else
    echo "Unrecongnized command line option. Valid options are \"true\" or \"false\""
fi

$ cat 99_master-dynamic-system-reserved-calc.yaml.in
...
    storage:
      files:
      - contents:
          source: data:text/plain;charset=utf-8;base64,@BASE64_CONTENT@
...
```
and replace `@BASE64_CONTENT@` with `cat dynamic-system-reserved-calc.sh | base64 -w0`

Comment (cfergeau) [reply 99_master-dynamic-system-reserved-calc.yaml:14]: I just noticed a `Unrecongnized` typo (`Unrecognized`)

**Review by cfergeau — COMMENTED** [review]
This adds a script to /usr/local/sbin/dynamic-system-reserved-calc.sh which will create `/etc/node-sizing.env` when it's called with `true`/`false`. I'm missing what is going to call it?
And the PR you link to documents
```
apiVersion: machineconfiguration.openshift.io/v1
kind: KubeletConfig
metadata:
  name: dynamic-node 
spec:
  autoSizingReserved: true 
  machineConfigPoolSelector:
    matchLabels:
      pools.operator.machineconfiguration.openshift.io/worker: ""
```
not clear why we are not making use of this?

File: 99_master-dynamic-system-reserved-calc.yaml (hunk 1 above)
Comment (guillaumerose) [reply 99_master-dynamic-system-reserved-calc.yaml:14]: +1 for this. Easier to search/understand.

**Comment by praveenkumar** [comment]
@guillaumerose @cfergeau need another round of review.

**Comment by cfergeau** [comment]
> @guillaumerose @cfergeau need another round of review.

I still have the same questions/lack of understanding :) https://github.com/code-ready/snc/pull/366#pullrequestreview-630204865

File: dynamic-system-reserved-calc.sh [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+    echo "SYSTEM_RESERVED_MEMORY=0Gi" >> ${NODE_SIZES_ENV}
+    echo "SYSTEM_RESERVED_CPU=0m" >> ${NODE_SIZES_ENV}
+}
+
+if [ $1 == "true" ]; then
+    dynamic_node_sizing
+elif [ $1 == "false" ]; then
+    static_node_sizing
+else
+    echo "Unrecongnized command line option. Valid options are \"true\" or \"false\""
```
Comment (cfergeau) [inline dynamic-system-reserved-calc.sh:24]: "Unrecognized" (there is one extra 'n')

**Comment by praveenkumar** [comment]
> This adds a script to /usr/local/sbin/dynamic-system-reserved-calc.sh which will create `/etc/node-sizing.env` when it's called with `true`/`false`. I'm missing what is going to call it?
> And the PR you link to documents

https://github.com/openshift/machine-config-operator/pull/2466/files is not the document link but it have all the implementation about dynamic allocation of cpu/mem according to configuration when user set it `true` otherwise it will make `1G` mem and `500m` cpu default https://github.com/openshift/machine-config-operator/pull/2466/files#diff-7f8d5dd4bf4dc8e795fe99a2931facec02ff7c017da2cfec7da60bf64f4f2e3aR86-R89 , The script detect node  memory and cpu and then allocate with different `if-else` clause.

**Comment by cfergeau** [comment]
Oh I did not realize this commit is overwriting a file which was introduced in this other PR. 
To answer my other question, this file is then used by:
```
name: kubelet-auto-node-size.service
enabled: true
contents: |
  [Unit]
  Description=Dynamically sets the system reserved for the kubelet
  Wants=network-online.target
  After=network-online.target ignition-firstboot-complete.service
  Before=kubelet.service crio.service
  [Service]
  # Need oneshot to delay kubelet
  Type=oneshot
  RemainAfterExit=yes
  EnvironmentFile=/etc/node-sizing-enabled.env
  ExecStart=/bin/bash /usr/local/sbin/dynamic-system-reserved-calc.sh ${NODE_SIZING_ENABLED}
  [Install]
  RequiredBy=kubelet.service
```
Still missing the full story though :) Are we going to enable dynamic node sizing (since you deal with it in the script)? Or do we expect to only keep using static node sizing? I'd argue the terminology is inverted btw, dynamic is no longer dynamic with that change, while static is also not static since we let the kubelet use whatever amount of memory it needs.

And more importantly, I'm worried we are changing something OpenShift consinders an internal implementation detail, and may change at any time. Is there any commitment from them/documentation/... that it's ok to replace that file with ours to do whatever we need?

**Comment by praveenkumar** [comment]
As per https://kubernetes.io/docs/tasks/administer-cluster/reserve-compute-resources/#system-reserved looks like we need to only reserve the resource for system daemons services like sshd, chronyd ..etc. So I run the crc without kubelet service to find out what is actual use of resources.

```
top - 11:06:09 up 1 min,  1 user,  load average: 0.08, 0.07, 0.02
Tasks: 170 total,   1 running, 168 sleeping,   0 stopped,   1 zombie
%Cpu(s):  0.1 us,  0.1 sy,  0.0 ni, 99.7 id,  0.0 wa,  0.1 hi,  0.1 si,  0.0 st
MiB Mem :   8970.2 total,   8328.7 free,    329.3 used,    312.1 buff/cache           0           0
```

So I think `350Mb` should be good enough and there is hardly any cpu uses so `100m` would be enough

File: 99_master-node-sizing-enabled-env.yaml.in [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+    machineconfiguration.openshift.io/role: master
+  name: 99-node-sizing-for-crc
+spec:
+  config:
+    ignition:
+      version: 3.2.0
+    storage:
+      files:
+      - contents:
+          source: data:text/plain;charset=utf-8;base64,${DYNAMIC_DATA}
```
Comment (cfergeau) [inline 99_master-node-sizing-enabled-env.yaml.in:14]: It seems this data can be inline? (looking at https://github.com/harche/machine-config-operator/blob/4fdea851cf8984722518895ed7b07b746ff09fa1/templates/common/_base/files/kubelet-auto-sizing.yaml )
```

contents:
  inline: |
    NODE_SIZING_ENABLED=false
    SYSTEM_RESERVED_MEMORY=350Mi
    SYSTEM_RESERVED_CPU=200m

File: snc.sh [hunk 4]
```diff
@@ -226,14 +226,15 @@ ${SSH} core@api.${CRC_VM_NAME}.${BASE_DOMAIN} -- 'sudo mv /home/core/kubeconfig
 
 # Export all manifests to the disk, modify them and use them in the CVO
 ${SSH} core@api.${CRC_VM_NAME}.${BASE_DOMAIN} -- 'sudo mkdir /opt/release-manifests/'
+retry ${OC} -n openshift-cluster-version get pods -o=name
```
Comment (cfergeau) [inline snc.sh:229]: log should be `Add missing retry for oc commands` (misplaced 'missing'). This is unrelated to this PR right? The CI issue would happen even without the previous changes?

Comment (praveenkumar) [reply snc.sh:229]: yes this is unrelated to the PR but CI actually failed twice on this one.

**Review by cfergeau — APPROVED** [review]

File: snc.sh [hunk 5]
```diff
@@ -224,14 +226,15 @@ ${SSH} core@api.${CRC_VM_NAME}.${BASE_DOMAIN} -- 'sudo mv /home/core/kubeconfig
...
```
Comment (guillaumerose) [inline snc.sh:229]: This line (229) can be removed. 
Otherwise /lgtm

**Pull request merged**